source/fuzz/transformation_add_loop_preheader.cpp

```diff
@@ -0,0 +1,218 @@
+    std::vector<uint32_t> phi_id) {
+  message_.set_loop_header_block(loop_header_block);
+  message_.set_fresh_id(fresh_id);
+  for (auto id : phi_id) {
+    message_.add_phi_id(id);
+  }
+}
+
+bool TransformationAddLoopPreheader::IsApplicable(
+    opt::IRContext* ir_context,
+    const TransformationContext& /* unused */) const {
+  // |message_.loop_header_block()| must be the id of a loop header block.
+  opt::BasicBlock* loop_header_block =
+      fuzzerutil::MaybeFindBlock(ir_context, message_.loop_header_block());
+  if (!loop_header_block || !loop_header_block->IsLoopHeader()) {
+    return false;
+  }
+
+  // The id for the preheader must actually be fresh.
+  std::set<uint32_t> used_ids;
+  if (!CheckIdIsFreshAndNotUsedByThisTransformation(message_.fresh_id(),
+                                                    ir_context, &used_ids)) {
+    return false;
+  }
+
+  // If the block only has one predecessor outside of the loop (and thus 2 in
+  // total), then no additional fresh ids are necessary.
+  if (ir_context->cfg()->preds(message_.loop_header_block()).size() == 2) {
+    return true;
+  }
+
+  // Count the number of OpPhi instructions.
+  int32_t num_phi_insts = 0;
+  loop_header_block->ForEachPhiInst(
+      [&num_phi_insts](opt::Instruction* /* unused */) { num_phi_insts++; });
+
+  // There must be enough fresh ids for the OpPhi instructions.
+  if (num_phi_insts > message_.phi_id_size()) {
+    return false;
+  }
+
+  // Check that the needed ids are fresh and distinct.
+  for (int32_t i = 0; i < num_phi_insts; i++) {
+    if (!CheckIdIsFreshAndNotUsedByThisTransformation(message_.phi_id(i),
+                                                      ir_context, &used_ids)) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+void TransformationAddLoopPreheader::Apply(
+    opt::IRContext* ir_context,
+    TransformationContext* /* transformation_context */) const {
+  // Find the loop header.
+  opt::BasicBlock* loop_header =
+      fuzzerutil::MaybeFindBlock(ir_context, message_.loop_header_block());
+
+  auto dominator_analysis =
+      ir_context->GetDominatorAnalysis(loop_header->GetParent());
+
+  uint32_t back_edge_block_id = 0;
+
+  // Update the branching instructions of the out-of-loop predecessors of the
+  // header. Set |back_edge_block_id| to be the id of the block from which the
+  // back edge starts.
+  ir_context->get_def_use_mgr()->ForEachUse(
+      loop_header->id(),
+      [this, &ir_context, &dominator_analysis, &loop_header,
+       &back_edge_block_id](opt::Instruction* use_inst, uint32_t use_index) {
+
+        if (dominator_analysis->Dominates(loop_header->GetLabelInst(),
+                                          use_inst)) {
+          // If |use_inst| is a branch instruction dominated by the header, the
+          // block containing it is the back-edge block.
+          if (use_inst->IsBranch()) {
+            back_edge_block_id = ir_context->get_instr_block(use_inst)->id();
```

Comment:
I suggest preceding this by an assertion that back_edge_block_id is 0, since there can only be one back-edge block.  This is an example of something that we don't really need to assert about - since we assume valid SPIR-V - but where I feel that the assertion would help someone reading the code understand that indeed this assignment should happen exactly once.

Reply:
Done